### PR TITLE
ENH: Support EAs in Series.unstack

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/23077

This prevents ExtensionArray-backed series from being converted to object-dtype in unstack.

The strategy is to do a dummy unstack on an ndarray of integers, which provides the `indices` to `take` later on. We then concat together at the end. This provided decent performance, and seems pretty maintainable in the long run.

I'll post some benchmarks later.

Do we want to do DataFrame.stack() in the same PR?

### CODE REVIEW

File: pandas/tests/extension/decimal/array.py
Comment (TomAugspurger) [inline pandas/tests/extension/decimal/array.py]: Series.equal doesn't consider `Series([np.nan]) == Series([Decimal('NaN')])`

**Review by jschendel — COMMENTED** [review]
A couple minor comments

File: doc/source/whatsnew/v0.24.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: really? what does this change for Categorical?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
> I think if we were designing that today, we wouldn't have allowed that.

I'm actually rethinking, this. Maybe we *would* want to allow this. It's a pretty clear statement of user intent, and I could easily imaging someone wanting to do something like "take, but fill missing values (-1) with "None" or "other").

**Comment by jorisvandenbossche** [comment]
Do you want to resolve the fill_value` question here, or leave for https://github.com/pandas-dev/pandas/issues/23296 ?

(as I mentioned there: I would preserve the dtype, which then means only allowing a fill_value that is NaN or within the categories)

**Comment by TomAugspurger** [comment]
We can ignore `fill_value` in this PR.

This is blocked by https://github.com/pandas-dev/pandas/issues/23296 for now.

File: doc/source/whatsnew/v0.24.0.txt
Comment (TomAugspurger) [reply doc/source/whatsnew/v0.24.0.txt]: Ah, I forget. Previously, we went internally went `Categorical -> object -> Categorical`. Now we avoid the conversion to categorical.

So the changes from 0.23.4 will be

1. Series[category].unstack() avoids a conversion to object
2. Series[Sparse].unstack is sparse (no intermediate conversion to dense)

Onces DatetimeTZ is an ExtensionArray, then we'll presumably preserve that as well. On 0.23.4, we convert to datetime64ns

```python
In [48]: index = pd.MultiIndex.from_tuples([('A', 0), ('A', 1), ('B', 1)])

In [49]: ser = pd.Series(pd.date_range('2000', periods=3, tz="US/Central"), index=index)

In [50]: ser.unstack().dtypes
Out[50]:
0    datetime64[ns]
1    datetime64[ns]
dtype: object
```

**Comment by TomAugspurger** [comment]
Apparently, `fill_value` was broken in 0.23.4 for Categortical

```python
In [51]: index = pd.MultiIndex.from_tuples([('A', 0), ('A', 1), ('B', 1)])

In [52]: ser = pd.Series(pd.Categorical(['a', 'b', 'a']), index=index)

In [53]: ser.unstack(fill_value='c')
Out[53]:
     0  1
A    a  b
B  NaN  a

In [54]: ser.unstack(fill_value='a')
Out[54]:
   0  1
A  a  b
B  a  a
```

We just silently didn't fill the `'d'`. So I guess we just raise an error there instead?

**Comment by jorisvandenbossche** [comment]
> So I guess we just raise an error there instead?

+1

File: pandas/tests/extension/base/reshaping.py
Comment (TomAugspurger) [reply pandas/tests/extension/base/reshaping.py]: Yes, you're correct.

**Review by jorisvandenbossche — COMMENTED** [review]
Does this now also work for unstacking a DataFrame with an EA column? If so, maybe add that to the test case?

File: doc/source/whatsnew/v0.24.0.txt
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: ok, this might be need a larger note then

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm, module a subsection in the docs

**Comment by jreback** [comment]
I also recall some issues that might be closed by this (haven't looked), but categorical reshaping might be a good place to start.

**Comment by TomAugspurger** [comment]
FYI, my comment about this changing behavior for Series / categorical wasn't correct. This will be backwards compat, we're just changing it so that sparse & categorical are no longer special-cased to work.

`DataFrame.unstack()` is proving to be a bit of work, but I should get that done today.

File: pandas/core/internals/blocks.py
Comment (TomAugspurger) [inline pandas/core/internals/blocks.py]: Ugh, so this is bad. There are several issues.

1. `_unstack_frame` partially applies `_Unstacker` and then applies it to the to the *blocks*` of a frame via `BlockManger._unstack`. For `Series[EA]`, we preempt the `_Unstacker` down in https://github.com/pandas-dev/pandas/pull/23284/files#diff-fef81b7e498e469973b2da18d19ff6f3R403.
2. modifying a partially applied function is probably a bad idea, but if I want to use a similar strategy, I need to extract the `fill_value` and then update it to use `-1`.
3. I do a dummy unstack on an empty array just to get the index length. This could maybe be avoided
4. After that, this is the same as NonConsolidatableMixin._untsack.

So consider this a WIP for now.

Comment (jorisvandenbossche) [reply pandas/core/internals/blocks.py]: Maybe you can change the `Block.unstack` signature to not pass through the partial object, but the actual object and keywords? So it less hacky to change something?

Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: At a glance, it looks like the `_Unstacker.__init__` is somewhat expensive. I suspect this was done originally for performance.

Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: Ah, but I could at least pass through the kwargs in addition to the partially applied _Unstacker.

Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: That would let us avoid the hacks with the `functools.partial`.

Comment (jorisvandenbossche) [reply pandas/core/internals/blocks.py]: But is the partial doing any caching? I would expect the `__init__` not yet been run in the partial, and after that it is called again for each block it seems? 
(but might be wrong, didn't look in detail)

Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: Right, Your correct.

**Comment by TomAugspurger** [comment]
Things are a bit better in 5902b5b.

**Comment by TomAugspurger** [comment]
Failure in https://travis-ci.org/pandas-dev/pandas/jobs/445842263 is unrelated (fixed in https://github.com/pandas-dev/pandas/pull/23323)

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py]: see my comment in other PR (about stack)

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py]: can you add a doc-string, i would just pass self (if you need to override) or make a module level function ; we don't use staticmethod too much

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py]: can you add a doc-string (or does it share)?

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py]: would not this generically work for all unstacking/ (e.g. what if you make this the super method)?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: The inherited one OK. I'm going to leave a comment explaining why we override.

File: pandas/core/internals/blocks.py
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: I think this is slower in general. It's necessarily slower for NumPy types, since you have to do the reshaping / unstack on the ndarray of positions anyway.

The hope is that the cost of the additional `take` per column is offset by not converting to an ndarray[object] in the first place.

**Comment by TomAugspurger** [comment]
Here are the results from that ASV

master + benchmark:

```
[  0.00%] ·· Benchmarking existing-py_Users_taugspurger_Envs_pandas-dev_bin_python3
[ 25.00%] ··· Running (reshape.Unstack.time_full_product--)..
[ 75.00%] ··· reshape.Unstack.time_full_product                                                                                                           ok
[ 75.00%] ··· ========== ==========
                param1
              ---------- ----------
                 int      73.0±2ms
               category   193±1ms
              ========== ==========

[100.00%] ··· reshape.Unstack.time_without_last_row                                                                                                       ok
[100.00%] ··· ========== ===========
                param1
              ---------- -----------
                 int      189±0.7ms
               category    509±5ms
              ========== ===========
```

and this PR

```
[  0.00%] ·· Benchmarking existing-py_Users_taugspurger_Envs_pandas-dev_bin_python3
[ 25.00%] ··· Running (reshape.Unstack.time_full_product--)..
[ 75.00%] ··· reshape.Unstack.time_full_product                                                                                                           ok
[ 75.00%] ··· ========== ============
                param1
              ---------- ------------
                 int      75.9±0.4ms
               category    192±4ms
              ========== ============

[100.00%] ··· reshape.Unstack.time_without_last_row                                                                                                       ok
[100.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      190±2ms
               category   515±2ms
              ========== =========

```


The int case should be unchanged from before. The only thing that should differ there is an extra `if is_extension_dtype` check.

The category one should be taking the new code path, but those timings are suspiciously close together that I'm going to double check.

**Comment by TomAugspurger** [comment]
Ugh, yeah I messed up the ASV. Forgot to convert to categorical.

```
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] ·· Benchmarking existing-py_Users_taugspurger_Envs_pandas-dev_bin_python3
[ 25.00%] ··· reshape.Unstack.time_full_product                                                                                       ok
[ 25.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      137±0ms
               category   17.2±0s
              ========== =========

[ 50.00%] ··· reshape.Unstack.time_without_last_row                                                                                   ok
[ 50.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      265±0ms
               category   17.5±0s
              ========== =========

```

Back to the drawing board...

**Comment by TomAugspurger** [comment]
Ah, I forgot to re-run the benchmark on master after converting to category. So we're actually faster!

Master:

```
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] ·· Benchmarking existing-py_Users_taugspurger_Envs_pandas-dev_bin_python3
[ 25.00%] ··· reshape.Unstack.time_full_product                                                                                                           ok
[ 25.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      136±0ms
               category   50.6±0s
              ========== =========

[ 50.00%] ··· reshape.Unstack.time_without_last_row                                                                                                       ok
[ 50.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      263±0ms
               category   49.6±0s
              ========== =========
```

This branch

```
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] ·· Benchmarking existing-py_Users_taugspurger_Envs_pandas-dev_bin_python3
[ 25.00%] ··· reshape.Unstack.time_full_product                                                                                                           ok
[ 25.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      141±0ms
               category   18.0±0s
              ========== =========

[ 50.00%] ··· reshape.Unstack.time_without_last_row                                                                                                       ok
[ 50.00%] ··· ========== =========
                param1
              ---------- ---------
                 int      257±0ms
               category   18.0±0s
              ========== =========

```

I'm going to update the benchmark to not take 18s though.

**Comment by TomAugspurger** [comment]
This should be good to go, assuming CI passes.

File: pandas/tests/extension/decimal/array.py
Comment (jreback) [inline pandas/tests/extension/decimal/array.py]: OT: how much testing do we have on this?

File: pandas/tests/extension/decimal/test_decimal.py
Comment (jreback) [inline pandas/tests/extension/decimal/test_decimal.py]: hmm, we *could* add this support I think though, create an issue?

File: pandas/tests/extension/decimal/test_decimal.py
Comment (jreback) [inline pandas/tests/extension/decimal/test_decimal.py]: where does this come up now? e.g. whats an example

**Review by jreback — APPROVED** [review]
happy to merge, just some questions about the decimal nan checking

File: pandas/tests/extension/decimal/test_decimal.py
Comment (TomAugspurger) [reply pandas/tests/extension/decimal/test_decimal.py]:
```python
In [21]: import pandas as pd

In [22]: import pandas.util.testing as tm

In [23]: from pandas.tests.extension.decimal import to_decimal

In [24]: ser = pd.Series(to_decimal(['1.0', 'NaN']))

In [25]: tm.assert_series_equal(ser.astype(object), ser.astype(object))
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-25-5356adf50e72> in <module>
----> 1 tm.assert_series_equal(ser.astype(object), ser.astype(object))

~/sandbox/pandas/pandas/util/testing.py in assert_series_equal(left, right, check_dtype, check_index_type, check_series_type, check_less_precise, check_names, check_exact, check_datetimelike_compat, check_categorical, obj)
   1293                                      check_less_precise=check_less_precise,
   1294                                      check_dtype=check_dtype,
-> 1295                                      obj='{obj}'.format(obj=obj))
   1296
   1297     # metadata comparison

~/sandbox/pandas/pandas/_libs/testing.pyx in pandas._libs.testing.assert_almost_equal()
     64
     65
---> 66 cpdef assert_almost_equal(a, b,
     67                           check_less_precise=False,
     68                           bint check_dtype=True,

~/sandbox/pandas/pandas/_libs/testing.pyx in pandas._libs.testing.assert_almost_equal()
    178             msg = '{0} values are different ({1} %)'.format(
    179                 obj, np.round(diff * 100.0 / na, 5))
--> 180             raise_assert_detail(obj, msg, lobj, robj)
    181
    182         return True

~/sandbox/pandas/pandas/util/testing.py in raise_assert_detail(obj, message, left, right, diff)
   1080         msg += "\n[diff]: {diff}".format(diff=diff)
   1081
-> 1082     raise AssertionError(msg)
   1083
   1084

AssertionError: Series are different

Series values are different (50.0 %)
[left]:  [1.0, NaN]
[right]: [1.0, NaN]
```

We do the astype(object) to build the expected.

File: pandas/tests/extension/decimal/test_decimal.py
Comment (TomAugspurger) [reply pandas/tests/extension/decimal/test_decimal.py]: Right now we have a pretty clear answer of "decimal isn't supported". I'm hesitant to even partially support it :)

File: pandas/tests/extension/decimal/array.py
Comment (TomAugspurger) [reply pandas/tests/extension/decimal/array.py]: Which bit specifically? We have a base test for `astype` to object and series.

I'm not actually sure why I changed this...

File: pandas/tests/extension/decimal/test_decimal.py
Comment (jreback) [reply pandas/tests/extension/decimal/test_decimal.py]: I think this is a bug, and would solve the aboe i think.

```
In [7]: ser.astype(object).isna()
Out[7]: 
0    False
1    False
dtype: bool
```

**Comment by TomAugspurger** [comment]
K, I'll fix the `pd.isna(Decimal('NaN'))` and revert the other changes.

**Comment by TomAugspurger** [comment]
Hmm, fixing `isna` for decimal seems to be expensive. `np.isnan` raises on decimal input, so we would need to search any object array for decimal elements, and call our isna on it. I don't think that's a good idea.

**Comment by jreback** [comment]
> Hmm, fixing isna for decimal seems to be expensive. np.isnan raises on decimal input, so we would need to search any object array for decimal elements, and call our isna on it. I don't think that's a good idea.

this would only be for object input, and can you all ``math.isnan``? (maybe instead)?

**Comment by TomAugspurger** [comment]
Ahh I thought we would have to pay the extra cost on string dtypes too, but it seems like those are handled before we get to a generic object dtype. This should be doable.

**Comment by TomAugspurger** [comment]
Well, I'm going back to -1 on support decimal here, unless we can find a better way than a basic isnstance.

```diff
diff --git a/doc/source/whatsnew/v0.24.0.txt b/doc/source/whatsnew/v0.24.0.txt
index f449ca532..c8c5db611 100644
--- a/doc/source/whatsnew/v0.24.0.txt
+++ b/doc/source/whatsnew/v0.24.0.txt
@@ -1227,6 +1227,7 @@ Missing
 - Bug in :func:`Series.hasnans` that could be incorrectly cached and return incorrect answers if null elements are introduced after an initial call (:issue:`19700`)
 - :func:`Series.isin` now treats all NaN-floats as equal also for `np.object`-dtype. This behavior is consistent with the behavior for float64 (:issue:`22119`)
 - :func:`unique` no longer mangles NaN-floats and the ``NaT``-object for `np.object`-dtype, i.e. ``NaT`` is no longer coerced to a NaN-value and is treated as a different entity. (:issue:`22295`)
+- :meth:`isna` now considers ``decimal.Decimal('NaN')`` a missing value (:issue:`23284`).
 
 
 MultiIndex
diff --git a/pandas/_libs/missing.pyx b/pandas/_libs/missing.pyx
index b87913592..4fa96f652 100644
--- a/pandas/_libs/missing.pyx
+++ b/pandas/_libs/missing.pyx
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 
 import cython
+import decimal
 from cython import Py_ssize_t
 
 import numpy as np
@@ -33,6 +34,8 @@ cdef inline bint _check_all_nulls(object val):
         res = get_datetime64_value(val) == NPY_NAT
     elif util.is_timedelta64_object(val):
         res = get_timedelta64_value(val) == NPY_NAT
+    elif isinstance(val, decimal.Decimal):
+        return val.is_nan()
     else:
         res = 0
     return res
@@ -71,6 +74,8 @@ cpdef bint checknull(object val):
         return get_timedelta64_value(val) == NPY_NAT
     elif util.is_array(val):
         return False
+    elif isinstance(val, decimal.Decimal):
+        return val.is_nan()
     else:
         return val is None or util.is_nan(val)
 
```

some timings

kind          | master | PR     | ratio
------------- | ------ | ------ | -----
scalar        | 821 ns | 920 ns | 1.12 (different result)
object array  | 1.0 ms | 2.4 ms | 2.4
decimal array | 1.0 ms | 3.6 ms | 3.6 (different result)

the object array is an object-dtype series with 20,000 elements. The decimal array is an object-dtype series with 20,000 decimal elements. I don't really care about the last one being 3.6x slower, since we're getting the correct result. I'm more concerned about the others.

**Comment by jreback** [comment]
this moves calls to python land. Try

```
elif hasattr(val, 'is_nan'):
    return val.is_nan()
```

**Comment by jreback** [comment]
hmm i guess just the additional check is causing this.

but a more general questions. *should* we even be checking for this in an ndarray object array at all? e.g. we don't do this for a random foo object. It must be a decimal array (in which case you just ask it .isnan).?

**Comment by jorisvandenbossche** [comment]
Shall we leave possible changes/support for decimal in the internals for another issue or PR? 
Because in some sense it is exactly the idea of an ExtensionArray that it can override the behaviour of `pd.isna` to do this correctly no?

**Comment by jreback** [comment]
> Shall we leave possible changes/support for decimal in the internals for another issue or PR?
Because in some sense it is exactly the idea of an ExtensionArray that it can override the behaviour of pd.isna to do this correctly no?

totally fine. @TomAugspurger can you xfail these tests rather than change them though. and create an issue to update.

**Comment by jorisvandenbossche** [comment]
> can you xfail these tests rather than change them though. and create an issue to update.

Isn't the current code fine? It's contained in `test_decimal.py`, I don't think it is a problem that there is some decimal specific code in that file?

File: pandas/tests/extension/decimal/test_decimal.py
Comment (jorisvandenbossche) [reply pandas/tests/extension/decimal/test_decimal.py]: @TomAugspurger actually, can you explain why the current code in master is not working fine? Why do you need to convert to object? Because before, there was already the calls to `isna` to check NaNs and non-NaNs separately.

Comment (jreback) [reply pandas/tests/extension/decimal/test_decimal.py]: this is my point. I think there is a bug somewhere here, e.g. isna is maybe not dispatching to the EA?

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/issues/23530 for isna(decimal).

Fixed the merge conflict.

File: pandas/tests/extension/decimal/test_decimal.py
Comment (TomAugspurger) [reply pandas/tests/extension/decimal/test_decimal.py]: To construct the expected in `test_untsack` base test, we do `ser.astype(object).untsack(level=level)`. That means we convert the Series[Decimal] to Series[object], which has the incorrect `isna`.

Comment (TomAugspurger) [reply pandas/tests/extension/decimal/test_decimal.py]: So it's all about creating the expected result. When we go to do the final assert that the values match, we do

```python
            result = result.astype(object)

            self.assert_frame_equal(result, expected)
```

but `self.assert_frame_equal` will say that `Series([Decimal('NaN')], dtype='object')` isn't equal to itself, since it doesn't consider that value NA.

Comment (jorisvandenbossche) [reply pandas/tests/extension/decimal/test_decimal.py]: Ah, OK, I somehow thought you were doing the `astype(object)` inside the assert testing machinery above, not in the actual expected result. Yes, that makes sense now. 

For me it is fine to keep this hack in here for now. In the end that is somehow the purpose of using the class instances for `assert_.._equal` so a specific EA can override it.

**Comment by TomAugspurger** [comment]
All green.

File: pandas/tests/extension/decimal/test_decimal.py
Comment (jreback) [reply pandas/tests/extension/decimal/test_decimal.py]: I think the ``.astype(object)`` for Decimal is incorrect.

```
In [5]: ser.values
Out[5]: DecimalArray(array([Decimal('1.0'), Decimal('NaN')], dtype=object))

In [6]: ser.astype(object)
Out[6]: 
0    1.0
1    NaN
dtype: object

In [7]: ser.astype(object).values
Out[7]: array([Decimal('1.0'), Decimal('NaN')], dtype=object)
```

on the these should be converted to ``np.nan`` and not ``Decimal('NaN')`` I think as this is just a numpy array.

Comment (TomAugspurger) [reply pandas/tests/extension/decimal/test_decimal.py]: object-dtype can store anything, including decimal objects. It'd be strange to only convert `Decimal("NaN")` to `np.nan`, and not `Decimal('1.0')` to 1.0, no?

Comment (jorisvandenbossche) [reply pandas/tests/extension/decimal/test_decimal.py]: Yeah, this seems correct to me. 

Going to merge, if we need to further discuss this, we can do that in another issue (it's not really related anymore with actually fixing unstack)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@TomAugspurger Thanks!

**Pull request merged**
